src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,268 @@
+/**
+ * @file gan.hpp
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_IMPL_HPP
+#define MLPACK_METHODS_ANN_GAN_IMPL_HPP
+
+#include "gan.hpp"
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/core/dists/gaussian_distribution.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::math;
+using namespace mlpack::distribution;
+
+namespace mlpack {
+namespace ann /** Restricted Boltzmann Machine.  */ {
+template<typename Generator, typename Discriminator, typename IntializerType>
+GenerativeAdversarialNetwork<Generator, Discriminator,IntializerType>
+::GenerativeAdversarialNetwork(arma::mat trainData,
+    arma::mat trainLables,
+    IntializerType initializeRule,
+    Generator& generator,
+    Discriminator& discriminator,
+    size_t batchSize,
+    size_t generatorInSize):
+    initializeRule(initializeRule),
+    generator(generator),
+    discriminator(discriminator),
+    trainGenerator(false),
+    batchSize(batchSize),
+    generatorInSize(generatorInSize),
+    reset(false)
+{
+  numFunctions = trainData.n_cols;
+  predictors = trainData;
+  responses = trainLables;
+  fakeData.set_size(trainData.n_rows, batchSize);
+}
+template<typename Generator, typename Discriminator, typename IntializerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator,IntializerType>
+::Reset()
+{
+  // Call the reset function of both the Generator and Discriminator Network
+  generator.ResetParameters();
+  discriminator.ResetParameters();
+  fakeData.set_size(predictors.n_rows, batchSize);
+  fakeLables.set_size(1, batchSize);
+  tempTrainData.set_size(predictors.n_rows, 2 * batchSize);
+  tempLabels.set_size(1, 2 * batchSize);
+  fakeData.zeros();
+  fakeLables.zeros();
+  parameter.set_size(generator.Parameters().n_rows + 
+      discriminator.Parameters().n_rows,
+      generator.Parameters().n_cols + generator.Parameters().n_cols);
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+  generator.Parameters() = arma::mat(parameter.memptr(),
```

Comment:
I think this line is incorrect since you can not change this pointer. I mean that the layers of the generator network still store data in the previous pointer. That leads to segfault. So, you can not change the pointer after `ResetParameters()`.